src/main/java/com/example/demo/model/Cart.java

```diff
@@ -0,0 +1,63 @@
+package com.example.demo.model;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+public class Cart {
+
+    private List<LineItem> lineItems;
+    private int totalQuantity;
+
+    public Cart(List<LineItem> lineItems) {
+        this.lineItems = new ArrayList<>(lineItems);
+        calculateTotalQuantity();
+    }
+
+    public List<LineItem> getLineItems() {
+        return Collections.unmodifiableList(lineItems);
+    }
+
+    public int getTotalQuantity() {
+        return totalQuantity;
+    }
+
+    public void addProduct(String productId, String productOption, int quantity) {
+
+        checkValidQuantity(quantity);
+
+        LineItem lineItem = findLineItem(productId, productOption);
+
+        if (lineItem != null) {
+            lineItem.addQuantity(quantity);
+            calculateTotalQuantity();
+            return;
+        }
+
+        lineItem = new LineItem(productId, productOption, quantity);
+        lineItems.add(lineItem);
+        calculateTotalQuantity();
+    }
+
+    public void checkValidQuantity(int quantity) {
+        if (totalQuantity + quantity> 20) {
+            throw new IllegalArgumentException("담을수 있는 수량을 초과했습니다.");
+        }
+    }
+
+    private void calculateTotalQuantity() {
+        this.totalQuantity = lineItems.stream()
+                .mapToInt(LineItem::getQuantity)
+                .sum();
+    }
+
+
+    private LineItem findLineItem(String productId, String productOption) {
+        return lineItems.stream()
+                .filter(i -> i.getProductId().equals(productId)
+                        && i.getProductOption().equals(productOption))
```

Comment:
동일 상품인지 검사하는 게 복잡해지면 어떻게 될까요? 옵션도 여러 개인 경우도 포함해서요. 예를 들어 구두를 판매한다고 할 때, 색상과 크기 두 가지 옵션을 가진 경우가 있을 수 있겠죠.
코드 레벨에서 뭔가 불편한 느낌이 들면 설계를 개선할 좋은 기회를 잡은 겁니다. 이런 건 LineItem에 위임하면 이런 모양이 되고, 앞으로 유지보수할 때 크게 도움이 될 겁니다.

```suggestion
                .filter(lineItem -> lineItem.isSameProduct(productId, productOption))
```